<!DOCTYPE html><html lang="en-US" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>spatial Transcription region deg genemodule heatmap GO analysis | zepoch's site</title><meta name="keywords" content="tech"><meta name="author" content="zepoch"><meta name="copyright" content="zepoch"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="After the spatial transcriptome is annotated manually, the regional annotation information of the spatial group is obtained. At this time, we want to obtain the functional annotation information of">
<meta property="og:type" content="article">
<meta property="og:title" content="spatial Transcription region deg genemodule heatmap GO analysis">
<meta property="og:url" content="https://www.zepoch.cc/2024/87416160.html">
<meta property="og:site_name" content="zepoch&#39;s site">
<meta property="og:description" content="After the spatial transcriptome is annotated manually, the regional annotation information of the spatial group is obtained. At this time, we want to obtain the functional annotation information of">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
<meta property="article:published_time" content="2024-01-03T02:50:03.000Z">
<meta property="article:modified_time" content="2024-01-03T02:59:38.953Z">
<meta property="article:author" content="zepoch">
<meta property="article:tag" content="tech">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"><link rel="shortcut icon" href="/img/favicon.ico"><link rel="canonical" href="https://www.zepoch.cc/2024/87416160"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//www.google-analytics.com" crossorigin=""/><link rel="preconnect" href="//www.clarity.ms"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><meta name="google-site-verification" content="AXddQAUftQD7NDjBUlQvw5LWu1x_A1uNpdVqI8c_3J0"/><meta name="baidu-site-verification" content="code-wVE0OwLLU2"/><meta name="bing_site_verification" content="154CD85FDB00EA195DB3F8F193CE78BE"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.css" media="print" onload="this.media='all'"><script async="async" src="https://www.googletagmanager.com/gtag/js?id=G-5K8YRSMKEE"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-5K8YRSMKEE');
</script><script>(function(c,l,a,r,i,t,y){
    c[a]=c[a]||function(){(c[a].q=c[a].q||[]).push(arguments)};
    t=l.createElement(r);t.async=1;t.src="https://www.clarity.ms/tag/"+i;
    y=l.getElementsByTagName(r)[0];y.parentNode.insertBefore(t,y);
})(window, document, "clarity", "script", "bernfipu7j");</script><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"search.xml","languages":{"hits_empty":"We didn't find any results for the search: ${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"prismjs","highlightCopy":true,"highlightLang":false,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: 'days',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery@2/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery@2/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: true,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'spatial Transcription region deg genemodule heatmap GO analysis',
  isPost: true,
  isHome: false,
  isHighlightShrink: undefined,
  isToc: false,
  postUpdate: '2024-01-03 10:59:38'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 5.4.2"><link rel="alternate" href="/atom.xml" title="zepoch's site" type="application/atom+xml">
</head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAMAAABOo35HAAAABGdBTUEAAK/INwWK6QAAABl0RVh0U29mdHdhcmUAQWRvYmUgSW1hZ2VSZWFkeXHJZTwAAAC9UExURVlZWdPT07KysmRkZIWFhfT09JmZmWZmZm9vb39/fxkZGUxMTDMzM3p6epCQkKamppubm729venp6cjIyN7e3tbW1s/Pz8LCwnx8fLS0tFZWVoiIiI+Pj6GhoeTk5Glpabu7u93d3evr66CgoJSUlKqqqsnJyeDg4Hd3d8PDw+Xl5bi4uNHR0dvb26Ojo6urq+fn51hYWDg4OCgoKHBwcK2traenp0FBQe7u7vHx8U5OTre3t8zMzHV1df///7GrnpQAAAA/dFJOU///////////////////////////////////////////////////////////////////////////////////AI4mfBcAAAUGSURBVHja7NoJb6M4GMZxY0NCD64kve/pMZ2d3Z297+X7f6zFNmBAMUXa6URl/q9UJSWPUPzrizFWRUlNLgEBWGCBBRZYYEEAFlhggQUWWBCABRZYYIEFFgRggQUWWGCBBQFYYIEFFlhgQQAWWGCBBRZYEIAFFlhggQUWBGCBBRZYYIEFAVhggQUWWGBBABZYYIEFFlgQgAUWWGCBBRYEYIEFFlhggQUBWGCBBRZYYEEAFlhggQUWWBCABRZYYIEFFgRggQUWWGCBBQFYYIEFFlhgQQAWWGCBBRZYEIAFFlhggQUWBGCBBRZYn6cCIcRXgvX/h9qcIVBqDdbEM8RCxGCB9QqXYRwHYDHBgwXWl8eKZKiESHI3Ba1kWs3fKixcaJUl1YyeBm7Ocq+yLItUiVBGnXxenSHJolIKEcwHq6ikbOX1YGVzQCTN8LPmSLreghUl9sN4Uw7yajMrLC0TZ1ImzqY6FEop0+pIaEN5HaoOxVuwEqFyc4I46uSlzOLqgxlh6UaR9l3VYWl9Fdoxb1Q90KJtu41pwwFW/WHhTtW8i7TafLCqRsk6bsGw63L9qurXRmuIlbT9lDQnlXU+nBFW1Q2qnZbDprWa2tjR90LZFqx1/+Td/HpGWLlrLDvIwTcx6dQ1Vrntbig68cDms3JwbA5Y1azs1ger6sNV/bbIw1jU81MvNAGrl58RVn8ozW+btF08iGFoAlYvP3csfVur1gJBEIA1uBmue5dhZDOyO2epbmgCVi8/I6x0MMHH9pjsTfBhNzQBq5uPZoQlB0uH3DZG4EZqQ26fL3sZq5uf09Ih6qw3i/pm6BZO0qZX7rrUS68Xsbr5ZE4rePMk08pk9aUZugfqppvs6AM1Acvlo/StP+6EbW06z8hJqxbYp2BZPQUnFsLsKuhQdaHqn5ewbF7KXIn0jWO5MqOQ7RaNLPtbNMmmhimj0GUmYLl8Gs0Lq4wyPbTu1l2QKqHSouzs3OlDIslW5SQsnY/NXmFplyNvEuuLV/Tau9BzwiraDUSwXmysztYWWNtL1psXeumgIrDGaqXvBfUuvtqUYI3V2t1wk1e2msFluJJm6zDJXv/fIfjPP7DAAgsssCiwwAILLLDAosACCyywwAKLAgsssMACC6zt9fDz/v75tyOB+98PD2+ORgKffjw4OP1uJPDxl+Xy8v1I4MPF3t7VNyOB4/vF4uzdzrG+39f1kz/w66Guv/yBvw90KX/gZKkr8Qf+2dOV+gNHC12/7RxrabD2/a31bLAO/a11YbAO/K21MFhLf2s9Gqw9f2vdGqzFu11jnVusE2/gxmI9eQOnFuvYG7i0WH7uK4t15w2cWazrXWP9a7H8f/bQYvm/6IPF+sF/pVssf19Ii/WH/0K2WH/uGuvEWC39gSdj9Twy+Rqri5EZx1gt/IE7Y/XoD1wbq9vd3w1PlufnD2OBp+ebm/uxwPHF6emnscDR4vLy41jg7vHq6sNY4Pr27OyYdRaLUrDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssL6u+k+AAQCR9eHtLKvLfwAAAABJRU5ErkJggg==" data-original="/img/avatar.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data is-center"><div class="data-item"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">72</div></a></div><div class="data-item"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">61</div></a></div><div class="data-item"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">7</div></a></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> about</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="not-top-img" id="page-header"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">zepoch's site</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> Search</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> about</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav></header><main class="layout" id="content-inner"><div id="post"><div id="post-info"><h1 class="post-title">spatial Transcription region deg genemodule heatmap GO analysis</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2024-01-03T02:50:03.000Z" title="Created 2024-01-03 10:50:03">2024-01-03</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2024-01-03T02:59:38.953Z" title="Updated 2024-01-03 10:59:38">2024-01-03</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">3.3k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>20min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="spatial Transcription region deg genemodule heatmap GO analysis"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div><article class="post-content" id="article-container"><p>After the spatial transcriptome is annotated manually, the regional
annotation information of the spatial group is obtained. At this time,
we want to obtain the functional annotation information of its
differentially expressed genes. There is no universal script here. I
only provide an idea. ,For reference.</p>
<p>Since most of our spatial group data is in the h5ad file format of
anndata, we need to convert it into the rds file format that can be read
by seurat. Here we use a convenient tool that we talked about in our
previous blog <a
href="https://www.zepoch.cc/2023/581620572.html">Interaction between rds
and h5ad Conversion</a>, the approximate code is as follows:</p>
<pre class="line-numbers language-R" data-language="R"><code class="language-R">sceasy::convertFormat(&#39;&#x2F;data&#x2F;*.h5ad&#39;, 
                      from&#x3D;&quot;anndata&quot;, 
                      to&#x3D;&quot;seurat&quot;, 
                      main_layer &#x3D; &#39;count&#39;,
                      outFile&#x3D;&#39;&#x2F;data&#x2F;*.rds&#39;)<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>So we got the corresponding rds file, and then we used the
findallmarkers function in seurat to screen differentially expressed
genes and remove some genes we don't need, such as mitochondrial genes,
ribosomal protein genes, hemoglobin genes, and cytoplasmic genes. ,
collagen genes, etc. Here we can filter through regular expressions and
sort avglog2fc to filter out the top 50 differentially expressed genes
of each group, and then calculate them with genemodule. What needs to be
noted here is that when I was calculating deg, I did not normalize the
data of the spatial group, because it is bin50 data. Of course, it is
not impossible to normalize it. You can decide whether you want to do
this. Carry out normalization; in addition, when calculating genemodule,
I chose to select the raw matrix and the matrix after normalization to
calculate genemodule. You can decide whether to normalize based on the
results. Although addgenemodule's The documentation states that the
input matrix is ​​a normalized matrix.</p>
<p>There are also some strange-looking codes in the script. The matrix
input to calculate deg and the matrix used to calculate genemodule can
be different. The reason is that we may capture certain positions on the
space group chip differently from other places, and we need to remove
them. At this time, we can select the matrix with certain areas removed
to calculate deg, and then use the normal chip to calculate
genemodule.</p>
<pre class="line-numbers language-R" data-language="R"><code class="language-R"># Rscript your_script.R path_to_rds_file.rds refined_pred_column_name path_to_output_csv_file.csv

# Read command line parameters
args &lt;- commandArgs(trailingOnly &#x3D; TRUE)

# Check that the correct number of parameters is provided
if (length(args) !&#x3D; 4) &#123;
  stop(&quot;Three parameters need to be provided：rds_deg_path, rds_module_path, refined_pred, def_save_path&quot;)
&#125;

# Parse command line parameters
rds_deg_path &lt;- args[1]
rds_module_path &lt;- args[2]
cluster &lt;- args[3]
save_path &lt;- args[4]

# Import required libraries
library(Seurat)
library(tidyverse)

# Read  data
rds &lt;- readRDS(rds_deg_path)

# Set the cell region label of the rds object
Idents(rds) &lt;- rds@meta.data[[cluster]]

# Find differentially expressed genes
df &lt;- FindAllMarkers(rds, only.pos &#x3D; TRUE)

# Remove genes starting with MT|HB|RP|AC|COL
#MT Mitochondria
# RP ribosomal protein
# HB hemoglobin
# AC Cytoplasm
# COL COLLAGEN
deg_save_path &#x3D; paste(save_path,&#39;deg&#x2F;deg.csv&#39;, sep &#x3D; &#39;&#39;)
write_csv(df, deg_save_path)

filtered_df &lt;- df[!grepl(&quot;^(MT|HB|RP|AC|COL)&quot;, df$gene), ]

# Create a list indexed by cluster, each cluster corresponds to a sub-data frame
cluster_list &lt;- split(filtered_df, filtered_df$cluster)
sort_and_select &lt;- function(df) &#123;
  sorted_df &lt;- df[order(df$avg_log2FC, decreasing &#x3D; TRUE), ]
  return(sorted_df)
&#125;

# Apply sorting and filtering functions to each subdataframe
sorted_dfs &lt;- lapply(cluster_list, sort_and_select)

# Merge all sorted sub-data frames into a new data frame
filtered_df &lt;- do.call(rbind, sorted_dfs)

deg_save_path &#x3D; paste(save_path,&#39;deg&#x2F;filtered_deg.csv&#39;, sep &#x3D; &#39;&#39;)
write_csv(filtered_df, deg_save_path)

# Create a list indexed by cluster, each cluster corresponds to a sub-data frame
cluster_list &lt;- split(filtered_df, filtered_df$cluster)

# Define a function that sorts a subdataframe by the avglog2fc column and filters the first 50 rows
sort_and_select &lt;- function(df) &#123;
  sorted_df &lt;- df[order(df$avg_log2FC, decreasing &#x3D; TRUE), ]
  if (nrow(sorted_df) &gt;&#x3D; 50) &#123;
    return(sorted_df[1:50, ])
  &#125; else &#123;
    return(sorted_df)
  &#125;
&#125;

# Apply sorting and filtering functions to each subdataframe
sorted_dfs &lt;- lapply(cluster_list, sort_and_select)

# Merge all sorted sub-data frames into a new data frame
new_dataframe &lt;- do.call(rbind, sorted_dfs)

# output final result
# print(new_dataframe)
deg_save_path &#x3D; paste(save_path,&#39;deg&#x2F;filtered_top50_deg.csv&#39;, sep &#x3D; &#39;&#39;)
write_csv(new_dataframe, deg_save_path)

# Save unique values ​​in cluster column into unique_cluster
unique_cluster &lt;- unique(new_dataframe$cluster)
# Loop over each unique cluster value
for (cluster_value in unique_cluster) &#123;
    print(cluster_value)
    subset_df &lt;- subset(filtered_df, cluster &#x3D;&#x3D; cluster_value)
    gene_list &lt;- subset_df$gene
    gene_list &#x3D; list(gene_list)
    gene_modules &#x3D; paste(cluster_value ,&#39;_module&#39;, sep &#x3D; &#39;&#39;)
    rds &lt;- readRDS(rds_module_path)
    rds &#x3D; AddModuleScore(rds, features&#x3D;gene_list,name&#x3D;&#39;gene_modules&#39;, nbin &#x3D; 12)
    gene_module_path &#x3D; paste(save_path, &#39;genemodules&#x2F;csv&#x2F;&#39;,cluster_value, &#39;_raw_genemodule.csv&#39;, sep &#x3D; &quot;&quot;)
    write.csv(rds@meta.data, gene_module_path)

    rds &#x3D; NormalizeData(rds)
    rds &#x3D; AddModuleScore(rds, features&#x3D;gene_list,name&#x3D;&#39;gene_modules&#39;, nbin &#x3D; 12)
    
    gene_module_path &#x3D; paste(save_path, &#39;genemodules&#x2F;csv&#x2F;&#39;,cluster_value, &#39;_lognor_genemodule.csv&#39;, sep &#x3D; &quot;&quot;)
    
    write.csv(rds@meta.data, gene_module_path)
&#125;<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>After calculating the information of genemodule, we need to draw it
on the spatial group to see whether these calculated deg are enriched in
the areas we divided. Scanpy is used here. The reason for using scanpy
is the spatial drawing function of scanpy. More versatile and
convenient.</p>
<pre class="line-numbers language-python" data-language="python"><code class="language-python"><span class="token comment"># python your_script.py --adata_file_path path_to_adata_file --csv_file_path path_to_csv_file --gene_modules module_name</span>

<span class="token keyword">import</span> pandas <span class="token keyword">as</span> pd
<span class="token keyword">import</span> scanpy <span class="token keyword">as</span> sc
<span class="token keyword">import</span> matplotlib<span class="token punctuation">.</span>pyplot <span class="token keyword">as</span> plt
<span class="token keyword">import</span> os
<span class="token keyword">import</span> argparse
<span class="token keyword">import</span> scanpy <span class="token keyword">as</span> sc

<span class="token comment"># Create parameter parser</span>
parser <span class="token operator">=</span> argparse<span class="token punctuation">.</span>ArgumentParser<span class="token punctuation">(</span><span class="token punctuation">)</span>
<span class="token comment"># Add command line options</span>
parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--adata_file_path'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Path to the AnnData file'</span><span class="token punctuation">)</span>
parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--save_path'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Name of the gene module column'</span><span class="token punctuation">)</span>

<span class="token comment"># Parse command line parameters</span>
args <span class="token operator">=</span> parser<span class="token punctuation">.</span>parse_args<span class="token punctuation">(</span><span class="token punctuation">)</span>

<span class="token comment"># Assign parameter values ​​to corresponding variables</span>
adata_file_path <span class="token operator">=</span> args<span class="token punctuation">.</span>adata_file_path
save_path <span class="token operator">=</span> args<span class="token punctuation">.</span>save_path

adata <span class="token operator">=</span> sc<span class="token punctuation">.</span>read<span class="token punctuation">(</span>adata_file_path<span class="token punctuation">)</span>

csv_path_dir <span class="token operator">=</span> os<span class="token punctuation">.</span>listdir<span class="token punctuation">(</span>save_path<span class="token operator">+</span><span class="token string">'genemodules/csv'</span><span class="token punctuation">)</span>
<span class="token keyword">print</span><span class="token punctuation">(</span>csv_path_dir<span class="token punctuation">)</span>

<span class="token keyword">for</span> i <span class="token keyword">in</span> csv_path_dir<span class="token punctuation">:</span>
    <span class="token keyword">if</span> <span class="token string">'genemodule'</span> <span class="token keyword">in</span> i<span class="token punctuation">:</span>
        <span class="token keyword">print</span><span class="token punctuation">(</span>i<span class="token punctuation">)</span>
        name <span class="token operator">=</span> i<span class="token punctuation">.</span>replace<span class="token punctuation">(</span><span class="token string">'.csv'</span><span class="token punctuation">,</span> <span class="token string">''</span><span class="token punctuation">)</span>
        gene_modules <span class="token operator">=</span> <span class="token string">'gene_modules1'</span> 

        csv <span class="token operator">=</span> pd<span class="token punctuation">.</span>read_csv<span class="token punctuation">(</span>save_path <span class="token operator">+</span><span class="token string">'genemodules/csv/'</span><span class="token operator">+</span> i<span class="token punctuation">)</span>
        gene_modules <span class="token operator">=</span> csv<span class="token punctuation">[</span>gene_modules<span class="token punctuation">]</span><span class="token punctuation">.</span>tolist<span class="token punctuation">(</span><span class="token punctuation">)</span>

        adata<span class="token punctuation">.</span>obs<span class="token punctuation">[</span>name<span class="token punctuation">]</span> <span class="token operator">=</span> gene_modules
        sc<span class="token punctuation">.</span>pl<span class="token punctuation">.</span>spatial<span class="token punctuation">(</span>adata<span class="token punctuation">,</span> color <span class="token operator">=</span> name<span class="token punctuation">,</span> spot_size <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">,</span> show <span class="token operator">=</span> <span class="token boolean">False</span><span class="token punctuation">,</span> frameon <span class="token operator">=</span> <span class="token boolean">False</span><span class="token punctuation">)</span>
        plt<span class="token punctuation">.</span>savefig<span class="token punctuation">(</span>save_path <span class="token operator">+</span> <span class="token string">'genemodules/fig/'</span><span class="token operator">+</span>i<span class="token punctuation">.</span>replace<span class="token punctuation">(</span><span class="token string">'.csv'</span><span class="token punctuation">,</span> <span class="token string">'.pdf'</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

    <span class="token keyword">else</span><span class="token punctuation">:</span>
        <span class="token keyword">continue</span>

<span class="token keyword">print</span><span class="token punctuation">(</span><span class="token string">'genemodule plot over!'</span><span class="token punctuation">)</span><span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>So far, deg has been calculated, genemodule has been calculated, and
drawing has been completed. Next, draw the heatmap. Here we use Python
for calculation and R language for drawing. We treat each area as a
psudo bulk. For each gene, calculate its average expression in each
psudo bulk, and then Save it as csv and then draw it using R
language.</p>
<pre class="line-numbers language-python" data-language="python"><code class="language-python"><span class="token keyword">import</span> pandas <span class="token keyword">as</span> pd
<span class="token keyword">import</span> anndata <span class="token keyword">as</span> ad
<span class="token keyword">import</span> os
<span class="token keyword">import</span> argparse
<span class="token keyword">import</span> scanpy <span class="token keyword">as</span> sc
parser <span class="token operator">=</span> argparse<span class="token punctuation">.</span>ArgumentParser<span class="token punctuation">(</span><span class="token punctuation">)</span>

parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--adata_file_path'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Path to the AnnData file'</span><span class="token punctuation">)</span>
parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--save_path'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Name of the gene module column'</span><span class="token punctuation">)</span>
parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--cluster'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Name of the cluster column'</span><span class="token punctuation">)</span>

args <span class="token operator">=</span> parser<span class="token punctuation">.</span>parse_args<span class="token punctuation">(</span><span class="token punctuation">)</span>

adata_file_path <span class="token operator">=</span> args<span class="token punctuation">.</span>adata_file_path
save_path <span class="token operator">=</span> args<span class="token punctuation">.</span>save_path
cluster <span class="token operator">=</span> args<span class="token punctuation">.</span>cluster

listdir <span class="token operator">=</span> os<span class="token punctuation">.</span>listdir<span class="token punctuation">(</span>save_path <span class="token operator">+</span> <span class="token string">'deg/'</span><span class="token punctuation">)</span>

<span class="token keyword">print</span><span class="token punctuation">(</span>listdir<span class="token punctuation">)</span>

<span class="token keyword">for</span> i <span class="token keyword">in</span> listdir<span class="token punctuation">:</span>
    <span class="token keyword">if</span> <span class="token string">'filtered'</span> <span class="token keyword">in</span> i<span class="token punctuation">:</span>
        adata <span class="token operator">=</span> ad<span class="token punctuation">.</span>read<span class="token punctuation">(</span>adata_file_path<span class="token punctuation">)</span>
        name <span class="token operator">=</span> i<span class="token punctuation">.</span>replace<span class="token punctuation">(</span><span class="token string">'.csv'</span><span class="token punctuation">,</span> <span class="token string">''</span><span class="token punctuation">)</span>
        <span class="token keyword">print</span><span class="token punctuation">(</span>name<span class="token punctuation">)</span>
        csv <span class="token operator">=</span> pd<span class="token punctuation">.</span>read_csv<span class="token punctuation">(</span>save_path <span class="token operator">+</span> <span class="token string">'deg/'</span> <span class="token operator">+</span> i<span class="token punctuation">)</span>
        deg_df_cluster <span class="token operator">=</span> <span class="token builtin">list</span><span class="token punctuation">(</span><span class="token builtin">set</span><span class="token punctuation">(</span>csv<span class="token punctuation">[</span><span class="token string">'cluster'</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
        all_genes_ <span class="token operator">=</span> <span class="token punctuation">[</span><span class="token punctuation">]</span>
        <span class="token keyword">for</span> j <span class="token keyword">in</span> deg_df_cluster<span class="token punctuation">:</span>
            temp <span class="token operator">=</span> csv<span class="token punctuation">[</span>csv<span class="token punctuation">[</span><span class="token string">'cluster'</span><span class="token punctuation">]</span> <span class="token operator">==</span>j<span class="token punctuation">]</span><span class="token punctuation">.</span>sort_values<span class="token punctuation">(</span><span class="token string">'avg_log2FC'</span><span class="token punctuation">,</span> ascending <span class="token operator">=</span> <span class="token boolean">False</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token number">50</span><span class="token punctuation">]</span><span class="token punctuation">.</span>gene<span class="token punctuation">.</span>tolist<span class="token punctuation">(</span><span class="token punctuation">)</span>
            all_genes_ <span class="token operator">+=</span> temp
        all_genes <span class="token operator">=</span> <span class="token punctuation">[</span><span class="token punctuation">]</span>
        
        <span class="token keyword">for</span> j <span class="token keyword">in</span> all_genes_<span class="token punctuation">:</span>
            <span class="token keyword">if</span> j <span class="token keyword">in</span> all_genes<span class="token punctuation">:</span>
                <span class="token keyword">continue</span>
            <span class="token keyword">else</span><span class="token punctuation">:</span>
                all_genes<span class="token punctuation">.</span>append<span class="token punctuation">(</span>j<span class="token punctuation">)</span>
        heatmap <span class="token operator">=</span> pd<span class="token punctuation">.</span>DataFrame<span class="token punctuation">(</span>index <span class="token operator">=</span> all_genes<span class="token punctuation">)</span>
        <span class="token keyword">for</span> j <span class="token keyword">in</span> deg_df_cluster<span class="token punctuation">:</span>
            
            adata_temp <span class="token operator">=</span> adata<span class="token punctuation">[</span>adata<span class="token punctuation">.</span>obs<span class="token punctuation">[</span>cluster<span class="token punctuation">]</span> <span class="token operator">==</span> j<span class="token punctuation">]</span>
            adata_temp <span class="token operator">=</span> adata_temp<span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token punctuation">,</span>all_genes<span class="token punctuation">]</span><span class="token punctuation">.</span>X<span class="token punctuation">.</span>todense<span class="token punctuation">(</span><span class="token punctuation">)</span>
            adata_temp <span class="token operator">=</span> pd<span class="token punctuation">.</span>DataFrame<span class="token punctuation">(</span>adata_temp<span class="token punctuation">,</span>columns <span class="token operator">=</span> all_genes<span class="token punctuation">)</span>
            heatmap<span class="token punctuation">[</span>j<span class="token punctuation">]</span> <span class="token operator">=</span> adata_temp<span class="token punctuation">.</span>mean<span class="token punctuation">(</span><span class="token punctuation">)</span>
        heatmap<span class="token punctuation">.</span>to_csv<span class="token punctuation">(</span>save_path <span class="token operator">+</span> <span class="token string">'heatmap/'</span> <span class="token operator">+</span> name <span class="token operator">+</span><span class="token string">'_raw_heatmap.csv'</span><span class="token punctuation">)</span>
        
        sc<span class="token punctuation">.</span>pp<span class="token punctuation">.</span>normalize_total<span class="token punctuation">(</span>adata<span class="token punctuation">,</span> target_sum<span class="token operator">=</span><span class="token number">1e4</span><span class="token punctuation">)</span>
        sc<span class="token punctuation">.</span>pp<span class="token punctuation">.</span>log1p<span class="token punctuation">(</span>adata<span class="token punctuation">)</span>
        adata <span class="token operator">=</span> adata<span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token punctuation">,</span>all_genes<span class="token punctuation">]</span>
        sc<span class="token punctuation">.</span>pp<span class="token punctuation">.</span>scale<span class="token punctuation">(</span>adata<span class="token punctuation">,</span> max_value<span class="token operator">=</span><span class="token number">10</span><span class="token punctuation">)</span>
        
        heatmap <span class="token operator">=</span> pd<span class="token punctuation">.</span>DataFrame<span class="token punctuation">(</span>index <span class="token operator">=</span> all_genes<span class="token punctuation">)</span>
        <span class="token keyword">for</span> j <span class="token keyword">in</span> deg_df_cluster<span class="token punctuation">:</span>
            
            adata_temp <span class="token operator">=</span> adata<span class="token punctuation">[</span>adata<span class="token punctuation">.</span>obs<span class="token punctuation">[</span>cluster<span class="token punctuation">]</span> <span class="token operator">==</span> j<span class="token punctuation">]</span>
            <span class="token comment"># print(adata_temp.X)</span>
            adata_temp <span class="token operator">=</span> adata_temp<span class="token punctuation">.</span>X
            adata_temp <span class="token operator">=</span> pd<span class="token punctuation">.</span>DataFrame<span class="token punctuation">(</span>adata_temp<span class="token punctuation">,</span>columns <span class="token operator">=</span> all_genes<span class="token punctuation">)</span>
            heatmap<span class="token punctuation">[</span>j<span class="token punctuation">]</span> <span class="token operator">=</span> adata_temp<span class="token punctuation">.</span>mean<span class="token punctuation">(</span><span class="token punctuation">)</span>
        heatmap<span class="token punctuation">.</span>to_csv<span class="token punctuation">(</span>save_path <span class="token operator">+</span> <span class="token string">'heatmap/'</span> <span class="token operator">+</span> name <span class="token operator">+</span><span class="token string">'_lognor_heatmap.csv'</span><span class="token punctuation">)</span>
        
        
    <span class="token keyword">else</span><span class="token punctuation">:</span>
        <span class="token keyword">continue</span><span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>The plotting function looks like this:</p>
<pre class="line-numbers language-R" data-language="R"><code class="language-R"># Rscript 1.R gene_list_value rds_path_value gene_modules_value gene_module_path_value gene_module_normalized_path_value

args &lt;- commandArgs(trailingOnly &#x3D; TRUE)

# Check that the correct number of parameters is provided
if (length(args) !&#x3D; 1) &#123;
  stop(&quot;A parameter needs to be provided：save_path&quot;)
&#125;

# Parse command line parameters
save_path &#x3D; args[1]

library(pheatmap)
library(ggplot2)
library(viridis)
library(tidyverse)

raw &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_raw_heatmap.csv&#39;, sep &#x3D; &#39;&#39;)

data &lt;- read.csv(raw, encoding&#x3D;&quot;UTF-8&quot;)
rownames(data) &#x3D; data$X
data$X &#x3D;NULL
mycolors &#x3D; viridis(11)

options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_raw_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,scale &#x3D; &quot;row&quot;,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_raw_scale_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


raw &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_lognor_heatmap.csv&#39;, sep &#x3D; &#39;&#39;)

data &lt;- read.csv(raw, encoding&#x3D;&quot;UTF-8&quot;)
rownames(data) &#x3D; data$X
data$X &#x3D;NULL
mycolors &#x3D; viridis(11)

options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_lognor_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,scale &#x3D; &quot;row&quot;,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_deg_lognor_scale_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


raw &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_raw_heatmap.csv&#39;, sep &#x3D; &#39;&#39;)

data &lt;- read.csv(raw, encoding&#x3D;&quot;UTF-8&quot;)
rownames(data) &#x3D; data$X
data$X &#x3D;NULL
mycolors &#x3D; viridis(11)

options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_raw_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,scale &#x3D; &quot;row&quot;,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_raw_scale_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


raw &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_lognor_heatmap.csv&#39;, sep &#x3D; &#39;&#39;)

data &lt;- read.csv(raw, encoding&#x3D;&quot;UTF-8&quot;)
rownames(data) &#x3D; data$X
data$X &#x3D;NULL
mycolors &#x3D; viridis(11)

options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_lognor_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)


options(repr.plot.width&#x3D;6, repr.plot.height&#x3D;20)
p &#x3D; pheatmap(data ,cluster_col &#x3D; FALSE,cluster_row &#x3D; FALSE,scale &#x3D; &quot;row&quot;,color &#x3D; colorRampPalette(mycolors)(100),)

raw_save_path &#x3D; paste(save_path, &#39;heatmap&#x2F;filtered_top50_deg_lognor_scale_heatmap.pdf&#39;, sep &#x3D; &#39;&#39;)
ggsave(raw_save_path, plot &#x3D; p, width&#x3D;6,height&#x3D;20,limitsize &#x3D; FALSE)
<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>The heatmap of the R function draws more than one graph, and multiple
graphs. You can choose the one you need. The last step is to perform GO
enrichment analysis. Here I use clusterProfiler to perform GO enrichment
analysis. The code here is also relatively simple. You can refer to it
yourself.</p>
<pre class="line-numbers language-R" data-language="R"><code class="language-R"># Rscript your_script.R path_to_rds_file.rds refined_pred_column_name path_to_output_csv_file.csv

args &lt;- commandArgs(trailingOnly &#x3D; TRUE)

if (length(args) !&#x3D; 1) &#123;
  stop(&quot;A parameter needs to be provided：save_path&quot;)
&#125;

save_path &lt;- args[1]

library(&quot;clusterProfiler&quot;)
#install.packages(&quot;BiocManager&quot;)
#BiocManager::install(&quot;clusterProfiler&quot;)
#BiocManager::install(&quot;org.Hs.eg.db&quot;)
library(org.Hs.eg.db)
library(DOSE)
library(enrichplot)
library(&quot;tidyverse&quot;)
#install.packages(&quot;tidyverse&quot;)
library(RColorBrewer)
library(tidyverse)


#### go
cal_go &#x3D; function(genelist, fig_name, csv_name) &#123;
  go &#x3D; enrichGO(genelist, 
                OrgDb &#x3D; org.Hs.eg.db,
                ont &#x3D; &#39;ALL&#39;,
                pAdjustMethod &#x3D; &quot;BH&quot;,
                pvalueCutoff &#x3D; 1,
                qvalueCutoff &#x3D; 1,
                keyType &#x3D; &quot;SYMBOL&quot;)
  go.count &#x3D; data.frame(go)
  go.count_sort &#x3D; go.count[order(go.count$pvalue, decreasing &#x3D; FALSE),]
  
  top10 &#x3D; function(i) &#123;
    k &#x3D; head(i[order(i$pvalue), ], 10);
    return(k)
  &#125;
  
  dp &#x3D; rbind(top10(go.count[go.count[&#39;ONTOLOGY&#39;] &#x3D;&#x3D; &quot;BP&quot;, ]),
             top10(go.count[go.count[&#39;ONTOLOGY&#39;] &#x3D;&#x3D; &quot;CC&quot;, ]),
             top10(go.count[go.count[&#39;ONTOLOGY&#39;] &#x3D;&#x3D; &quot;MF&quot;, ]))
  d1 &#x3D; dp[which(dp[, 1] &#x3D;&#x3D; &quot;BP&quot;), ]
  d2 &#x3D; dp[which(dp[, 1] &#x3D;&#x3D; &quot;CC&quot;), ]
  d3 &#x3D; dp[which(dp[, 1] &#x3D;&#x3D; &quot;MF&quot;), ]
  
  d_l &#x3D; rbind(d1, d2, d3)
  
  mt &#x3D; d_l[, c(1, 3, 10)]
  
  colnames(mt)&lt;-c(&quot;Type&quot;,&quot;name&quot;,&quot;number&quot;)
  mt$name &lt;- factor(mt$name, levels &#x3D; rev(mt$name))
  
  ggplot(data&#x3D;mt, aes(x&#x3D;name,y&#x3D;number,fill&#x3D;Type))+ theme_bw()+
    geom_bar(stat &#x3D; &quot;identity&quot;, position&#x3D;position_dodge())+coord_flip()+xlab(&quot;GO term&quot;) + ylab(&quot;Number of genes&quot;)+
    scale_fill_brewer(palette&#x3D;&quot;Set2&quot;)+
    theme(panel.border &#x3D; element_blank(),
          plot.title &#x3D; element_text(face &#x3D; &quot;bold&quot;,size &#x3D; 12,vjust &#x3D;1),
          axis.title &#x3D; element_text(face &#x3D; &quot;bold&quot;,size &#x3D; 12, vjust &#x3D; 0.3),
          axis.text &#x3D; element_text(size &#x3D; 8, vjust &#x3D; 0.8), 
          axis.line &#x3D; element_line(), 
          panel.grid &#x3D; element_blank(),
          legend.position&#x3D;c(0.8,0.75),
          legend.key.size &#x3D; unit(4, &quot;mm&quot;))
  ggsave(fig_name, height &#x3D; 5, width &#x3D; 10)
  #write.csv(go.count_sort, csv_name, row.names &#x3D; F, quote &#x3D; F)
  go.count_sort %&gt;% write_csv(csv_name)
&#125;


dataframe_path &#x3D; paste(save_path, &#39;deg&#x2F;&#39;, &#39;filtered_deg.csv&#39;,sep &#x3D; &#39;&#39;)

dataframe &#x3D; read_csv(dataframe_path, col_names &#x3D; FALSE)
dataframe &lt;- dataframe[-1, ]

unique_clusters  &lt;- unique(dataframe$X6)

for (cluster_value  in unique_clusters ) &#123;
    print(cluster_value)
    subset_df &lt;- dataframe[dataframe$X6 &#x3D;&#x3D; cluster_value, ]
    
    genelist &#x3D; subset_df$X7
    genelist &#x3D; genelist[!is.na(genelist)]
    genelist &#x3D; genelist[-1]
    
    csv_result &#x3D; paste(save_path, &#39;go&#x2F;raw_results&#x2F;&#39;, cluster_value,&#39;.csv&#39;, sep &#x3D; &#39;&#39;)
    pdf_result &#x3D; paste(save_path, &#39;go&#x2F;raw_results&#x2F;&#39;, cluster_value,&#39;.pdf&#39;, sep &#x3D; &#39;&#39;)
    cal_go(genelist, pdf_result, csv_result)
&#125;


dataframe_path &#x3D; paste(save_path, &#39;deg&#x2F;&#39;, &#39;filtered_top50_deg.csv&#39;,sep &#x3D; &#39;&#39;)

dataframe &#x3D; read_csv(dataframe_path, col_names &#x3D; FALSE)
dataframe &lt;- dataframe[-1, ]

unique_cluster &lt;- unique(dataframe$X6)

for (cluster_value in unique_cluster) &#123;
    subset_df &lt;- dataframe[dataframe$X6 &#x3D;&#x3D; cluster_value, ]
    genelist &#x3D; subset_df$X7
    genelist &#x3D; genelist[!is.na(genelist)]
    genelist &#x3D; genelist[-1]
    csv_result &#x3D; paste(save_path, &#39;go&#x2F;raw_results&#x2F;&#39;, cluster_value,&#39;_top50.csv&#39;, sep &#x3D; &#39;&#39;)
    pdf_result &#x3D; paste(save_path, &#39;go&#x2F;raw_results&#x2F;&#39;, cluster_value,&#39;_top50.pdf&#39;, sep &#x3D; &#39;&#39;)
    cal_go(genelist, pdf_result, csv_result)
&#125;<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>This code can give a rough reference. Sometimes we only want
BP-related data. We can also read the table generated above and redraw
it. The code is as follows:</p>
<pre class="line-numbers language-python" data-language="python"><code class="language-python"><span class="token keyword">import</span> pandas <span class="token keyword">as</span> pd
<span class="token keyword">import</span> matplotlib<span class="token punctuation">.</span>pyplot <span class="token keyword">as</span> plt
<span class="token keyword">import</span> matplotlib<span class="token punctuation">.</span>ticker <span class="token keyword">as</span> ticker
<span class="token keyword">import</span> os
<span class="token keyword">import</span> argparse

parser <span class="token operator">=</span> argparse<span class="token punctuation">.</span>ArgumentParser<span class="token punctuation">(</span><span class="token punctuation">)</span>

parser<span class="token punctuation">.</span>add_argument<span class="token punctuation">(</span><span class="token string">'--save_path'</span><span class="token punctuation">,</span> <span class="token builtin">type</span><span class="token operator">=</span><span class="token builtin">str</span><span class="token punctuation">,</span> <span class="token builtin">help</span><span class="token operator">=</span><span class="token string">'Name of the gene module column'</span><span class="token punctuation">)</span>

args <span class="token operator">=</span> parser<span class="token punctuation">.</span>parse_args<span class="token punctuation">(</span><span class="token punctuation">)</span>

save_path <span class="token operator">=</span> args<span class="token punctuation">.</span>save_path

csv_file_dir_path <span class="token operator">=</span> save_path <span class="token operator">+</span> <span class="token string">'go/raw_results/'</span>
save_dir_path <span class="token operator">=</span> save_path <span class="token operator">+</span> <span class="token string">'go/filtered_results/'</span>
csv_file_dir <span class="token operator">=</span> os<span class="token punctuation">.</span>listdir<span class="token punctuation">(</span>csv_file_dir_path<span class="token punctuation">)</span>
csv_file <span class="token operator">=</span> <span class="token punctuation">[</span><span class="token punctuation">]</span>
<span class="token keyword">for</span> i <span class="token keyword">in</span> csv_file_dir<span class="token punctuation">:</span>
    <span class="token keyword">if</span> <span class="token string">'.csv'</span> <span class="token keyword">in</span> i<span class="token punctuation">:</span>
        csv_file<span class="token punctuation">.</span>append<span class="token punctuation">(</span>i<span class="token punctuation">)</span>

<span class="token keyword">def</span> <span class="token function">go_bp_plot</span><span class="token punctuation">(</span>csv_file_path<span class="token punctuation">,</span> save_path<span class="token punctuation">)</span><span class="token punctuation">:</span>
    csv <span class="token operator">=</span> pd<span class="token punctuation">.</span>read_csv<span class="token punctuation">(</span>csv_file_path<span class="token punctuation">)</span>
    csv <span class="token operator">=</span> csv<span class="token punctuation">[</span>csv<span class="token punctuation">[</span><span class="token string">'ONTOLOGY'</span><span class="token punctuation">]</span><span class="token operator">==</span><span class="token string">'BP'</span><span class="token punctuation">]</span>
    <span class="token comment"># csv = csv.drop_duplicates(subset=['geneID'])</span>
    csv<span class="token punctuation">.</span>index <span class="token operator">=</span> <span class="token builtin">range</span><span class="token punctuation">(</span><span class="token builtin">len</span><span class="token punctuation">(</span>csv<span class="token punctuation">)</span><span class="token punctuation">)</span>


    plt<span class="token punctuation">.</span>rcParams<span class="token punctuation">[</span><span class="token string">'font.size'</span><span class="token punctuation">]</span> <span class="token operator">=</span> <span class="token number">8</span>  

    ax <span class="token operator">=</span> csv<span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token number">10</span><span class="token punctuation">]</span><span class="token punctuation">.</span>plot<span class="token punctuation">.</span>barh<span class="token punctuation">(</span>y <span class="token operator">=</span> <span class="token string">'Count'</span><span class="token punctuation">,</span> x <span class="token operator">=</span> <span class="token string">'Description'</span><span class="token punctuation">,</span> width<span class="token operator">=</span><span class="token number">0.92</span><span class="token punctuation">,</span> color<span class="token operator">=</span><span class="token string">'#66c2a5'</span><span class="token punctuation">)</span>

    ax<span class="token punctuation">.</span>set_xlabel<span class="token punctuation">(</span><span class="token string">'Number of genes'</span><span class="token punctuation">,</span> fontsize<span class="token operator">=</span><span class="token string">'12'</span><span class="token punctuation">)</span>
    ax<span class="token punctuation">.</span>set_ylabel<span class="token punctuation">(</span><span class="token string">'GO term'</span><span class="token punctuation">,</span> fontsize<span class="token operator">=</span><span class="token string">'12'</span><span class="token punctuation">)</span>

    ax<span class="token punctuation">.</span>set_yticklabels<span class="token punctuation">(</span><span class="token punctuation">[</span><span class="token punctuation">]</span><span class="token punctuation">)</span> <span class="token comment"># Remove y-axis tick labels</span>

    <span class="token comment"># Add Description labels on top of each bar</span>
    max_v <span class="token operator">=</span> <span class="token builtin">max</span><span class="token punctuation">(</span>csv<span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token number">10</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'Count'</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
    <span class="token keyword">for</span> i<span class="token punctuation">,</span> v <span class="token keyword">in</span> <span class="token builtin">enumerate</span><span class="token punctuation">(</span>csv<span class="token punctuation">[</span><span class="token punctuation">:</span><span class="token number">10</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'Count'</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">:</span>
        Description <span class="token operator">=</span> csv<span class="token punctuation">[</span><span class="token string">'Description'</span><span class="token punctuation">]</span><span class="token punctuation">[</span>i<span class="token punctuation">]</span>
        ax<span class="token punctuation">.</span>text<span class="token punctuation">(</span><span class="token number">0.025</span><span class="token operator">*</span>max_v<span class="token punctuation">,</span> i<span class="token punctuation">,</span> Description<span class="token punctuation">,</span> color<span class="token operator">=</span><span class="token string">'black'</span><span class="token punctuation">,</span> va<span class="token operator">=</span><span class="token string">'center'</span><span class="token punctuation">)</span>
    <span class="token comment"># csv['Description'][index]</span>

    ax<span class="token punctuation">.</span>spines<span class="token punctuation">[</span><span class="token string">'top'</span><span class="token punctuation">]</span><span class="token punctuation">.</span>set_visible<span class="token punctuation">(</span><span class="token boolean">False</span><span class="token punctuation">)</span>
    ax<span class="token punctuation">.</span>spines<span class="token punctuation">[</span><span class="token string">'right'</span><span class="token punctuation">]</span><span class="token punctuation">.</span>set_visible<span class="token punctuation">(</span><span class="token boolean">False</span><span class="token punctuation">)</span>

    <span class="token comment"># Set x-axis tick locator</span>
    ax<span class="token punctuation">.</span>xaxis<span class="token punctuation">.</span>set_major_locator<span class="token punctuation">(</span>ticker<span class="token punctuation">.</span>MaxNLocator<span class="token punctuation">(</span>integer<span class="token operator">=</span><span class="token boolean">True</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

    <span class="token comment"># Set x-axis tick formatter</span>
    ax<span class="token punctuation">.</span>xaxis<span class="token punctuation">.</span>set_major_formatter<span class="token punctuation">(</span>ticker<span class="token punctuation">.</span>FormatStrFormatter<span class="token punctuation">(</span><span class="token string">'%d'</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

    ax<span class="token punctuation">.</span>legend<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">.</span>remove<span class="token punctuation">(</span><span class="token punctuation">)</span>
    plt<span class="token punctuation">.</span>savefig<span class="token punctuation">(</span>save_path<span class="token punctuation">,</span> <span class="token builtin">format</span><span class="token operator">=</span><span class="token string">'pdf'</span><span class="token punctuation">)</span>
    plt<span class="token punctuation">.</span>close<span class="token punctuation">(</span><span class="token punctuation">)</span>

    
<span class="token keyword">for</span> i <span class="token keyword">in</span> csv_file<span class="token punctuation">:</span>
    name <span class="token operator">=</span> i<span class="token punctuation">.</span>split<span class="token punctuation">(</span><span class="token string">'.'</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">0</span><span class="token punctuation">]</span>
    go_bp_plot<span class="token punctuation">(</span>csv_file_dir_path <span class="token operator">+</span> name <span class="token operator">+</span> <span class="token string">'.csv'</span><span class="token punctuation">,</span> save_dir_path <span class="token operator">+</span> name <span class="token operator">+</span> <span class="token string">'.pdf'</span><span class="token punctuation">)</span><span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>In general, we have implemented the calculation of deg of spatial
group data, the calculation and drawing of genemodule, the calculation
of the expression amount of each gene in each spatial annotation region
and the drawing of heatmap, as well as the GO analysis of each
region.</p>
<p>Finally, there is a simple sh script to run these 6 python or R
scripts in a unified manner, as shown below:</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"><span class="token comment"># Save address of all files</span>
<span class="token assign-left variable">save_path</span><span class="token operator">=</span><span class="token string">"/data/gw21_scp_0103/"</span>
<span class="token comment"># plot genemodule heatmap 's h5ad path</span>
<span class="token assign-left variable">h5ad_path</span><span class="token operator">=</span><span class="token string">"/data/*_0.01.h5ad"</span>
<span class="token comment"># deg's rds path</span>
<span class="token assign-left variable">rds_deg_path</span><span class="token operator">=</span><span class="token string">"/data/*_0103.rds"</span>
<span class="token comment"># module rds path</span>
<span class="token assign-left variable">rds_module_path</span><span class="token operator">=</span><span class="token string">"/data/*_0102.rds"</span>
<span class="token comment"># Annotation information for brain regions</span>
<span class="token assign-left variable">region</span><span class="token operator">=</span><span class="token string">'Ribbon'</span>

<span class="token assign-left variable">folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">""</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">deg_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"deg/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$deg_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"deg_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$deg_folder</span>"</span>
<span class="token keyword">fi</span>


<span class="token assign-left variable">genemodules_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"genemodules/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"genemodules_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">genemodules_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"genemodules/csv/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"genemodules_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">genemodules_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"genemodules/fig/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"genemodules_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$genemodules_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">heatmap_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"heatmap/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$heatmap_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"genemodule_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$heatmap_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">go_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"go/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$go_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"go_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$go_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">go_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"go/raw_results/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$go_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"go_raw_results_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$go_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token assign-left variable">go_folder</span><span class="token operator">=</span><span class="token variable">$&#123;save_path&#125;</span><span class="token string">"go/filtered_results/"</span>

<span class="token keyword">if</span> <span class="token punctuation">[</span> -d <span class="token string">"<span class="token variable">$go_folder</span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"go_filtered_results_folder already exists!"</span>
<span class="token keyword">else</span>
    <span class="token function">mkdir</span> <span class="token string">"<span class="token variable">$go_folder</span>"</span>
<span class="token keyword">fi</span>

<span class="token comment"># deg and genemodule calculations</span>
Rscript /data/work/17.Script/addgenemodule/1.R <span class="token variable">$&#123;rds_deg_path&#125;</span> <span class="token variable">$&#123;rds_module_path&#125;</span> <span class="token variable">$&#123;region&#125;</span> <span class="token variable">$&#123;save_path&#125;</span>
<span class="token comment"># genemodule plot</span>
python /data/work/17.Script/addgenemodule/2.py --adata_file_path <span class="token variable">$&#123;h5ad_path&#125;</span> --save_path <span class="token variable">$&#123;save_path&#125;</span>
<span class="token comment"># heatmap calculations</span>
python /data/work/17.Script/addgenemodule/3.py --adata_file_path <span class="token variable">$&#123;h5ad_path&#125;</span> --save_path <span class="token variable">$&#123;save_path&#125;</span> --cluster <span class="token variable">$&#123;region&#125;</span>
<span class="token comment"># heatmap plot</span>
Rscript /data/work/17.Script/addgenemodule/4.R <span class="token variable">$&#123;save_path&#125;</span>
<span class="token comment"># go calculations</span>
Rscript /data/work/17.Script/addgenemodule/5.R <span class="token variable">$&#123;save_path&#125;</span>
<span class="token comment"># go plot</span>
python /data/work/17.Script/addgenemodule/6.py --save_path <span class="token variable">$&#123;save_path&#125;</span><span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>that's all.</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">zepoch</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="https://www.zepoch.cc/2024/87416160.html">https://www.zepoch.cc/2024/87416160.html</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/tech/">tech</a></div><div class="post_share"><div class="social-share" data-image="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/37434737.html"><img class="prev-cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">计算某个区域中细胞的复杂性(2)</div></div></a></div><div class="next-post pull-right"><a href="/2024/397295361.html"><img class="next-cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">空间组deg genemodule heatmap go分析</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2023/2669081178.html" title="Calculate the complexity of cells in a region"><img class="cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-12-29</div><div class="title">Calculate the complexity of cells in a region</div></div></a></div><div><a href="/2024/4104152713.html" title="各区域不同细胞类型的统计"><img class="cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-01-02</div><div class="title">各区域不同细胞类型的统计</div></div></a></div><div><a href="/2024/397295361.html" title="空间组deg genemodule heatmap go分析"><img class="cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-01-03</div><div class="title">空间组deg genemodule heatmap go分析</div></div></a></div><div><a href="/2023/959813387.html" title="计算某个区域中细胞的复杂性"><img class="cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-12-29</div><div class="title">计算某个区域中细胞的复杂性</div></div></a></div><div><a href="/2024/37434737.html" title="计算某个区域中细胞的复杂性(2)"><img class="cover" src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-01-03</div><div class="title">计算某个区域中细胞的复杂性(2)</div></div></a></div></div></div><hr/><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> Comment</span></div></div><div class="comment-wrap"><div><div class="vcomment" id="vcomment"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAMAAABOo35HAAAABGdBTUEAAK/INwWK6QAAABl0RVh0U29mdHdhcmUAQWRvYmUgSW1hZ2VSZWFkeXHJZTwAAAC9UExURVlZWdPT07KysmRkZIWFhfT09JmZmWZmZm9vb39/fxkZGUxMTDMzM3p6epCQkKamppubm729venp6cjIyN7e3tbW1s/Pz8LCwnx8fLS0tFZWVoiIiI+Pj6GhoeTk5Glpabu7u93d3evr66CgoJSUlKqqqsnJyeDg4Hd3d8PDw+Xl5bi4uNHR0dvb26Ojo6urq+fn51hYWDg4OCgoKHBwcK2traenp0FBQe7u7vHx8U5OTre3t8zMzHV1df///7GrnpQAAAA/dFJOU///////////////////////////////////////////////////////////////////////////////////AI4mfBcAAAUGSURBVHja7NoJb6M4GMZxY0NCD64kve/pMZ2d3Z297+X7f6zFNmBAMUXa6URl/q9UJSWPUPzrizFWRUlNLgEBWGCBBRZYYEEAFlhggQUWWBCABRZYYIEFFgRggQUWWGCBBQFYYIEFFlhgQQAWWGCBBRZYEIAFFlhggQUWBGCBBRZYYIEFAVhggQUWWGBBABZYYIEFFlgQgAUWWGCBBRYEYIEFFlhggQUBWGCBBRZYYEEAFlhggQUWWBCABRZYYIEFFgRggQUWWGCBBQFYYIEFFlhgQQAWWGCBBRZYEIAFFlhggQUWBGCBBRZYn6cCIcRXgvX/h9qcIVBqDdbEM8RCxGCB9QqXYRwHYDHBgwXWl8eKZKiESHI3Ba1kWs3fKixcaJUl1YyeBm7Ocq+yLItUiVBGnXxenSHJolIKEcwHq6ikbOX1YGVzQCTN8LPmSLreghUl9sN4Uw7yajMrLC0TZ1ImzqY6FEop0+pIaEN5HaoOxVuwEqFyc4I46uSlzOLqgxlh6UaR9l3VYWl9Fdoxb1Q90KJtu41pwwFW/WHhTtW8i7TafLCqRsk6bsGw63L9qurXRmuIlbT9lDQnlXU+nBFW1Q2qnZbDprWa2tjR90LZFqx1/+Td/HpGWLlrLDvIwTcx6dQ1Vrntbig68cDms3JwbA5Y1azs1ger6sNV/bbIw1jU81MvNAGrl58RVn8ozW+btF08iGFoAlYvP3csfVur1gJBEIA1uBmue5dhZDOyO2epbmgCVi8/I6x0MMHH9pjsTfBhNzQBq5uPZoQlB0uH3DZG4EZqQ26fL3sZq5uf09Ih6qw3i/pm6BZO0qZX7rrUS68Xsbr5ZE4rePMk08pk9aUZugfqppvs6AM1Acvlo/StP+6EbW06z8hJqxbYp2BZPQUnFsLsKuhQdaHqn5ewbF7KXIn0jWO5MqOQ7RaNLPtbNMmmhimj0GUmYLl8Gs0Lq4wyPbTu1l2QKqHSouzs3OlDIslW5SQsnY/NXmFplyNvEuuLV/Tau9BzwiraDUSwXmysztYWWNtL1psXeumgIrDGaqXvBfUuvtqUYI3V2t1wk1e2msFluJJm6zDJXv/fIfjPP7DAAgsssCiwwAILLLDAosACCyywwAKLAgsssMACC6zt9fDz/v75tyOB+98PD2+ORgKffjw4OP1uJPDxl+Xy8v1I4MPF3t7VNyOB4/vF4uzdzrG+39f1kz/w66Guv/yBvw90KX/gZKkr8Qf+2dOV+gNHC12/7RxrabD2/a31bLAO/a11YbAO/K21MFhLf2s9Gqw9f2vdGqzFu11jnVusE2/gxmI9eQOnFuvYG7i0WH7uK4t15w2cWazrXWP9a7H8f/bQYvm/6IPF+sF/pVssf19Ii/WH/0K2WH/uGuvEWC39gSdj9Twy+Rqri5EZx1gt/IE7Y/XoD1wbq9vd3w1PlufnD2OBp+ebm/uxwPHF6emnscDR4vLy41jg7vHq6sNY4Pr27OyYdRaLUrDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssMCiwAILLLDAAosCCyywwAILLAossMACCyywKLDAAgsssL6u+k+AAQCR9eHtLKvLfwAAAABJRU5ErkJggg==" data-original="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">zepoch</div><div class="author-info__description">A full-stack engineer who is improving every day</div></div><div class="card-info-data is-center"><div class="card-info-data-item"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">72</div></a></div><div class="card-info-data-item"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">61</div></a></div><div class="card-info-data-item"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">7</div></a></div></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/zepoch"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/zepoch" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="mailto:zhotoa@gmail.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/37434737.html" title="计算某个区域中细胞的复杂性(2)">计算某个区域中细胞的复杂性(2)</a><time datetime="2024-01-03T13:25:23.000Z" title="Created 2024-01-03 21:25:23">2024-01-03</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/87416160.html" title="spatial Transcription region deg genemodule heatmap GO analysis">spatial Transcription region deg genemodule heatmap GO analysis</a><time datetime="2024-01-03T02:50:03.000Z" title="Created 2024-01-03 10:50:03">2024-01-03</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/397295361.html" title="空间组deg genemodule heatmap go分析">空间组deg genemodule heatmap go分析</a><time datetime="2024-01-03T02:19:37.000Z" title="Created 2024-01-03 10:19:37">2024-01-03</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/4104152713.html" title="各区域不同细胞类型的统计">各区域不同细胞类型的统计</a><time datetime="2024-01-02T12:23:57.000Z" title="Created 2024-01-02 20:23:57">2024-01-02</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/179608785.html" title="月度消费总结1">月度消费总结1</a><time datetime="2024-01-01T11:27:42.000Z" title="Created 2024-01-01 19:27:42">2024-01-01</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 - 2024 By zepoch</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><a id="to_comment" href="#post-comment" title="Scroll To Comments"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">Local search</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  Loading the Database</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="Search for Posts" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.js"></script><script src="https://jsd.012700.xyz/npm/instant.page@latest/instantpage.min.js" type="module"></script><script src="https://jsd.012700.xyz/npm/vanilla-lazyload@latest/dist/lazyload.iife.min.js"></script><script>function panguFn () {
  if (typeof pangu === 'object') pangu.autoSpacingPage()
  else {
    getScript('https://cdn.jsdelivr.net/npm/pangu/dist/browser/pangu.min.js')
      .then(() => {
        pangu.autoSpacingPage()
      })
  }
}

function panguInit () {
  if (false){
    GLOBAL_CONFIG_SITE.isPost && panguFn()
  } else {
    panguFn()
  }
}

document.addEventListener('DOMContentLoaded', panguInit)</script><script src="/js/search/local-search.js"></script><div class="js-pjax"><script>function loadValine () {
  function initValine () {
    const valine = new Valine(Object.assign({
      el: '#vcomment',
      appId: 'WspkUq7TuRDz5tk68Ry6TSa1-gzGzoHsz',
      appKey: 'xz4SYcXtJNP1hns8iBFAlhHT',
      avatar: 'monsterid',
      serverURLs: '',
      emojiMaps: "",
      path: window.location.pathname,
      visitor: true
    }, null))
  }

  if (typeof Valine === 'function') initValine() 
  else getScript('https://cdn.jsdelivr.net/npm/valine/dist/Valine.min.js').then(initValine)
}

if ('Valine' === 'Valine' || !true) {
  if (true) btf.loadComment(document.getElementById('vcomment'),loadValine)
  else setTimeout(loadValine, 0)
} else {
  function loadOtherComment () {
    loadValine()
  }
}</script></div><script async data-pjax src="https://cdn.bootcdn.net/ajax/libs/busuanzi/2.3.0/bsz.pure.mini.js"></script></div><script>
            window.imageLazyLoadSetting = {
                isSPA: true,
                preloadRatio: 1,
                processImages: null,
            };
        </script><script>window.addEventListener("load",function(){var t=/\.(gif|jpg|jpeg|tiff|png)$/i,r=/^data:image\/[a-z]+;base64,/;Array.prototype.slice.call(document.querySelectorAll("img[data-original]")).forEach(function(a){var e=a.parentNode;"A"===e.tagName&&(e.href.match(t)||e.href.match(r))&&(e.href=a.dataset.original)})});</script><script>!function(n){n.imageLazyLoadSetting.processImages=o;var e=n.imageLazyLoadSetting.isSPA,i=n.imageLazyLoadSetting.preloadRatio||1,r=Array.prototype.slice.call(document.querySelectorAll("img[data-original]"));function o(){e&&(r=Array.prototype.slice.call(document.querySelectorAll("img[data-original]")));for(var t,a=0;a<r.length;a++)0<=(t=(t=r[a]).getBoundingClientRect()).bottom&&0<=t.left&&t.top<=(n.innerHeight*i||document.documentElement.clientHeight*i)&&function(){var t,e,n,i,o=r[a];t=o,e=function(){r=r.filter(function(t){return o!==t})},n=new Image,i=t.getAttribute("data-original"),n.onload=function(){t.src=i,e&&e()},t.src!==i&&(n.src=i)}()}o(),n.addEventListener("scroll",function(){var t,e;t=o,e=n,clearTimeout(t.tId),t.tId=setTimeout(function(){t.call(e)},500)})}(this);</script></body></html>